dynamicio/io/resource.py

```diff
@@ -0,0 +1,97 @@
+from abc import abstractmethod
+from copy import deepcopy
+from pathlib import Path
+from typing import Callable, List, Optional, Type
+
+import pandas as pd
+from pandera import SchemaModel
+from pydantic import BaseModel
+from uhura import Readable, Writable
+
+from dynamicio.inject import InjectionError, inject
+from dynamicio.io.serde import BaseSerde, PickleSerde
+
+
+def create_schema_validator(schema) -> Callable[[pd.DataFrame], pd.DataFrame]:
+    def validate_schema(df: pd.DataFrame):
+        return schema.validate(df)
+
+    return validate_schema
+
+
+class BaseResource(BaseModel, Readable[pd.DataFrame], Writable[pd.DataFrame]):  # Cacheable kaputt, ABC kaputt
+    """Base class for all resources.
+
+    :injectables: List of attributes that can be injected with format strings.
+    :test_path (optional): Path to the test data. If set, the resource will be read from and written to this path.
+    :pa_schema (optional): Pandera schema to validate the resource. If set, the resource will be validated before writing and after reading.
+    """
+
+    injectables: List[str]
+    test_path: Optional[Path] = None
+    pa_schema: Optional[Type[SchemaModel]] = None
+
+    def inject(self, **kwargs) -> "BaseResource":
+        # copy object
```

Comment:
nit: I think that it may be worth adding a docstring describing what this method is meant to do and maybe also showing an example. As, for most people not familiar with the context it may seem rather abstract

Reply:
`maybe also showing an example` I saw that there is a test that could be used to understand this but it may be nicer for the next developer to have something here too

Reply:
Yes, definitely should exist. These things can be done when overhauling the readme / docs (one of the next steps). 

I think there is some auto-generated doc stuff but not sure how it works at the moment. The various doc hooks have some quite stringent requirements which might need some overhauling too.

Reply:
Adding something to this method for now.